File: gtest/test_addmv.cpp
#include <gtest/gtest.h>
#include "addmv.h"  // Include the header for your custom addmv class

// Define a struct for test parameters
struct AddMVTestParams {
    torch::Tensor input;
    torch::Tensor matrix;
    torch::Tensor vector;
    double beta;
    double alpha;
};

// Define the CustomaddmvTest class
class CustomAddMVTest : public ::testing::TestWithParam<AddMVTestParams> {
protected:
    custom_namespace::AddMV* custom_addmv;

    // Setup function to initialize the test fixture
    void SetUp() override {
        custom_addmv = new custom_namespace::AddMV();
    }

    // Tear down function to clean up after tests
    void TearDown() override {
        delete custom_addmv;
    }
};

// Instantiate parameterized tests with different input parameters
INSTANTIATE_TEST_SUITE_P(
    CustomAddMVTests,
    CustomAddMVTest,
    ::testing::Values(
                // Basic valid case
        AddMVTestParams{torch::tensor({1.0, 2.0, 3.0}), torch::rand({3, 2}), torch::rand({2}), 1.0, 1.0, },
        
        // Zero input tensor
        AddMVTestParams{torch::tensor({0.0, 0.0, 0.0}), torch::rand({3, 2}), torch::rand({2}), 0.0, 1.0, },

        // Negative values in input
        AddMVTestParams{torch::tensor({-1.0, 2.0, 3.0}), torch::rand({3, 2}), torch::rand({2}), 1.0, 1.0, },

        // Different scaling factors
        AddMVTestParams{torch::tensor({1.0, 2.0, 3.0}), torch::rand({3, 3}), torch::rand({3}), 2.0, 1.0, },

        // Matrices with small values
        AddMVTestParams{torch::tensor({0.5, 0.5}), torch::rand({2, 2}), torch::rand({2}), 1.0, 2.0, },

        // Single element tensor
        AddMVTestParams{torch::tensor({1.0}), torch::rand({1, 2}), torch::rand({2}), 1.0, 1.0, },

        // Empty tensor case
        AddMVTestParams{torch::tensor({}), torch::rand({0, 2}), torch::rand({2}), 1.0, 1.0, },

        // Random tensor with uniform distribution
        AddMVTestParams{torch::rand({100}), torch::rand({100, 50}), torch::rand({50}), 1.0, 1.0, },

        // Random tensor with normal distribution 
        AddMVTestParams{torch::randn({10}), torch::randn({10, 50}), torch::randn({50}), 1.0, 1.0}

    )
);

// Parameterized test for Custom AddMV function
TEST_P(CustomAddMVTest, ParameterizedAddMV) {
    AddMVTestParams params = GetParam();

    torch::Tensor result = custom_addmv->compute_addmv(params.input, params.matrix, params.vector, params.beta, params.alpha);
    
    // Expected result: manually perform the matrix multiplication and addition
    torch::Tensor expected = torch::addmv(params.input, params.matrix, params.vector, params.beta, params.alpha);

    // Compare results with a tolerance of 1e-5
    ASSERT_TRUE(torch::allclose(result, expected, 1e-5));
}

File: test/test_addmv.py
import torch, pytest
import custom_module

torch.manual_seed(2)


@pytest.mark.parametrize(
    "input, matrix, vector",
    (
        (torch.randn([2]), torch.randn([2, 3]), torch.randn([3])),
        (torch.randn([4]), torch.randn([4, 5]), torch.randn([5])),
        (torch.ones(3), torch.zeros(3, 2), torch.ones(2)),  # Zero tensor and ones
        (torch.ones(2), torch.ones(2, 3), torch.ones(3)),  # Ones tensor
    ),
)
@pytest.mark.parametrize("alpha", [-55.0, -33.0, 0.0, 1.0, 11.0, 22.0])
@pytest.mark.parametrize("beta", [-11.0, -22.0, 0.0, 1.0, 22.0, 11.0])
def test_AddMV(input, matrix, vector, alpha, beta):

    # Create an instance of AddMV
    ch = custom_module.AddMV()

    # Perform custom AddMV
    custom_AddMV = ch.compute_addmv(input, matrix, vector, alpha=alpha, beta=beta)

    torch_AddMV = torch.addmv(input, matrix, vector, alpha=alpha, beta=beta)

    # Assert that the custom result matches the expected output
    assert torch.allclose(
        custom_AddMV, torch_AddMV, rtol=1e1, atol=1e1
    ), f"custom_AddMV({input}, {matrix}, {vector}) != torch_AddMV({input}, {matrix}, {vector})"
